### PR TITLE
CSLI-15 contract state

### PR DESCRIPTION
**Pull request by samingle**
Known issues:
• scroll positioning on "Search" button could be better, especially for short lists
• ran out of time to get csli.json reading out of the home directory as a fallback

### CODE REVIEW

File: src/battle-station.tsx [hunk 1]
```diff
@@ -14,7 +14,7 @@ import * as jq from 'node-jq'
 import { TxHashes } from './panes/transactions/tx-hashes'
 import { TxDetails } from './panes/transactions/tx-details'
 import BlockDetailsPane from './panes/blocks/block-details'
-import { Debuggah } from "./panes/debug/debug";
+import Debuggah from "./panes/debug/debug";
```
Comment (kvnm) [inline src/battle-station.tsx:17]: phantom semicolon

File: src/contracts.tsx [hunk 2]
```diff
@@ -0,0 +1,75 @@
+import React, { useEffect, useState, useCallback } from 'react'
+import blessed from 'blessed'
+import { CosmWasmClient } from "@cosmjs/cosmwasm-stargate";
```
Comment (kvnm) [inline src/contracts.tsx:3]: ;

File: src/contracts.tsx [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+import ContractState from './panes/contracts/contract-state'
+
+interface ContractsProps {
+  screen: blessed.Widgets.Screen
+  client: CosmWasmClient
+  wss: Server
+}
+
+const Contracts: React.FC<ContractsProps> = ({screen, client, wss }) => {
+  const [selectedContractAddress, setSelectedContractAddress] = useState<string | undefined>(undefined )
```
Comment (kvnm) [inline src/contracts.tsx:22]: )

File: src/contracts.tsx [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  const [contractState, setContractState] = useState<any>(undefined)
+  const [query, setQuery] = useState<any>(undefined)
+  
+  useEffect(() => {
```
Comment (kvnm) [inline src/contracts.tsx:26]: what's this about? leftover debug?

Comment (samingle) [reply src/contracts.tsx:26]: ah yup

File: src/panes/contracts/query-options.tsx [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+        top: 0,
+        width: "100%",
+        height: "100%",
+      })
+      let offset = 0
+      let elementFocusRefs: string[] = []
+      const tabIndexOffset = 1.0 / (2.0 + (query.mandatory?.length ?? 0) + (query.optional?.length ?? 0))
+      let tabIndex = 3.0
+  
+      const buildInputUI = (parameter: any, index: number): {parameter: any, textbox: blessed.Widgets.TextboxElement} => {
```
Comment (kvnm) [inline src/panes/contracts/query-options.tsx:76]: this is great

File: src/panes/transactions/tx-details.tsx [hunk 6]
```diff
@@ -18,10 +22,27 @@ export const TxDetails: React.FC<TxDetailsProps> = ({ txData, isFocused }) => {
       }
     }
   }
+  const ref = useRef<blessed.Widgets.BoxElement>(null)
+  useEffect(() => {
```
Comment (kvnm) [inline src/panes/transactions/tx-details.tsx:26]: feels like we should make a ticket to handle focus for everything in a higher-level component so we don't have to do this in every child

File: src/tab-view.tsx [hunk 7]
```diff
@@ -0,0 +1,132 @@
+import React, { useState, useCallback, useEffect, useRef } from 'react'
+import blessed from 'blessed'
+import { CosmWasmClient } from "@cosmjs/cosmwasm-stargate";
```
Comment (kvnm) [inline src/tab-view.tsx:3]: ;

File: src/panes/transactions/tx-details.tsx (hunk 6 above)
Comment (samingle) [reply src/panes/transactions/tx-details.tsx:26]: the problem is you need a reference to the element in order to focus it

File: src/utils/react.ts [hunk 8]
```diff
@@ -0,0 +1,9 @@
+import {useRef, useEffect} from 'react';
```
Comment (kvnm) [inline src/utils/react.ts:1]: ; * 4

File: src/panes/transactions/tx-details.tsx (hunk 6 above)
Comment (kvnm) [reply src/panes/transactions/tx-details.tsx:26]: think this problem goes away when we leave react?

Comment (samingle) [reply src/panes/transactions/tx-details.tsx:26]: nope, you still have to call `.focus()` on the actual element that wants to be focused regardless

Comment (samingle) [reply src/panes/transactions/tx-details.tsx:26]: but there will be a lot less boilerplate without react

**Comment by kvnm** [comment]
Screen Shot 2022-08-19 at 3 00 23 PM

werks

**Pull request merged**